src/ui/components/Listbox/Listbox.js

```diff
@@ -0,0 +1,130 @@
+ui.class.Listbox = class Listbox extends HTMLElement {
+	constructor(props) {
+		super();
+		const self = this;
+
+		let m_parentTag;
+		let m_listboxContainerTag;
+		let m_props;
+		let m_listboxData;
+
+		this.addToList = addToList;
+		this.destroy = destroyListbox;
+		this.replaceList = replaceList;
+		this.setup = setup;
+
+		if (props) {
+			setup(props);
+		}
+
+		async function setup(props) {
+			await configure(props);
+			await ui.utils.installUIComponent({ self, m_parentTag, m_props });
+			await setupDOM();
+			setupComplete();
+		}
+
+		function setupDOM() {
+			return new Promise((resolve) => {
+
+				const tagContainerConfig = {
+					class: m_props.css.container,
+					prop: m_props.prop,
+				};
+
+				m_listboxContainerTag = ui.d.createTag(m_props.tags.container, tagContainerConfig);
+				self.classList.add(m_props.css.self);
+
+				installItems(m_listboxData).then(() => {
+					ui.d.append(self, m_listboxContainerTag);
+					resolve();
+				});
+
+			});
+		}
+
+		function installItems(data) {
+			return new Promise((resolve) => {
+				const itemPromises = data.map((itemData) => {
+					return processItem(itemData);
+				});
+				Promise.all(itemPromises).then(resolve);
+			});
+		}
+
+		function processItem(itemData) {
+			return new Promise((resolve) => {
+				const itemConfig = {
+					class: m_props.css.item,
+					text: itemData[m_props.displayKey]
+				}
+
+				let itemTag = ui.d.createTag(m_props.tags.item, itemConfig);
+				m_listboxContainerTag.appendChild(itemTag);
+
+				setupItemEventListeners(itemTag, itemData).then(resolve);
+			});
+		}
+
+		function addToList(data) {
+			m_listboxData = data;
+			installItems(m_listboxData).then();
+		}
+
+		function replaceList(data) {
+			clearData();
+			m_listboxData = data;
+			installItems(m_listboxData).then();
+		}
+
+		function clearData() {
+			m_listboxContainerTag.innerHTML = "";
+		}
+
+		function destroyListbox() {
+			ui.d.remove(self);
+		}
+
+		function setupItemEventListeners(item, data) {
+			return new Promise((resolve) => {
```

Comment:
Move this function setupItemEventListeners a bit higher up, grouped with the other Promise functions. I like to have all Promise functiones together at the beginning of the class, in order if possible.